File: src/lib/pty.ts
/**
 * PTY wrapper — event-based replacement for tauri-pty.
 *
 * Purpose: Provides the same IPty interface as tauri-pty but uses Tauri events
 * (push-based) instead of invoke polling. Implements real pause/resume for
 * flow control. Two-phase startup eliminates data-loss race conditions.
 *
 * Key decisions:
 *   - Constructor returns immediately; the actual spawn is async via `_ready`.
 *   - Output flows over a binary `tauri::ipc::Channel` (WI-1.1, ADR-T1): the
 *     reader thread sends `InvokeResponseBody::Raw(bytes)`, which the webview
 *     receives as an `ArrayBuffer` — NOT a JSON number array. This is ~3.66x
 *     less wire data and orders of magnitude less encode/decode CPU than the
 *     old `pty:data:` event path (see dev-docs/grills/terminal/). The Channel
 *     is point-to-point, so output is no longer broadcast to every window.
 *   - The data Channel's `onmessage` is wired BEFORE `pty_start` is invoked, so
 *     the reader cannot emit before we are listening — no data-loss race
 *     (this replaces the old two-phase listen-then-start dance for output).
 *   - The exit signal stays a plain `pty:exit:{pid}` event (low-frequency).
 *   - `pause()` and `resume()` are real Tauri commands (not stubs), enabling
 *     the watermark-based flow control in spawnPty.ts.
 *   - `kill()` eagerly cleans up event listeners and guards against mid-setup
 *     races via a `_destroyed` flag.
 *   - `pty_close` frees the Rust-side session (FDs/channels/child handle). It
 *     runs from the exit handler on natural exit; but because `kill()` (and the
 *     mid-setup guard) tear down the exit listener first, those paths call
 *     `pty_close` directly so the session is never leaked (#974).
 *
 * @coordinates-with src-tauri/src/pty.rs — Rust backend commands and events
 * @coordinates-with components/Terminal/spawnPty.ts — consumes this wrapper
 * @module lib/pty
 */

import { invoke, Channel } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import { ptyWarn, terminalLog } from "@/utils/debug";
import { errorMessage } from "@/utils/errorMessage";

// ---------------------------------------------------------------------------
// Public types — match the tauri-pty interface that spawnPty.ts expects
// ---------------------------------------------------------------------------

export interface IDisposable {
  dispose(): void;
}

export type IEvent<T> = (listener: (data: T) => void) => IDisposable;

export interface IPtyExitEvent {
  exitCode: number;
}

export interface IPtySpawnOptions {
  name?: string;
  cols?: number;
  rows?: number;
  cwd?: string;
  env?: Record<string, string>;
}

export interface IPty {
  readonly pid: number;
  cols: number;
  rows: number;
  readonly onData: IEvent<Uint8Array | number[]>;
  readonly onExit: IEvent<IPtyExitEvent>;
  write(data: string): void;
  resize(columns: number, rows: number): void;
  kill(): void;
  pause(): void;
  resume(): void;
}

// ---------------------------------------------------------------------------
// EventEmitter — minimal pub/sub
// ---------------------------------------------------------------------------

class EventEmitter<T> {
  private _listeners: Array<(data: T) => void> = [];

  get event(): IEvent<T> {
    return (listener) => {
      this._listeners.push(listener);
      return {
        dispose: () => {
          const idx = this._listeners.indexOf(listener);
          if (idx >= 0) this._listeners.splice(idx, 1);
        },
      };
    };
  }

  fire(data: T): void {
    for (const fn of [...this._listeners]) {
      fn(data);
    }
  }
}

// ---------------------------------------------------------------------------
// VMarkPty — the concrete implementation
// ---------------------------------------------------------------------------

class VMarkPty implements IPty {
  private _pid = 0;
  cols: number;
  rows: number;

  private _onData = new EventEmitter<Uint8Array | number[]>();
  private _onExit = new EventEmitter<IPtyExitEvent>();
  private _ready: Promise<void>;
  private _dataChannel: Channel<ArrayBuffer | Uint8Array | number[]> | null = null;
  private _unlistenExit: UnlistenFn | null = null;
  private _destroyed = false;

  get pid(): number {
    return this._pid;
  }

  constructor(file: string, args: string[], opts?: IPtySpawnOptions) {
    this.cols = opts?.cols ?? 80;
    this.rows = opts?.rows ?? 24;
    this._ready = this._setup(file, args, opts);
  }

  get onData(): IEvent<Uint8Array | number[]> {
    return this._onData.event;
  }
  get onExit(): IEvent<IPtyExitEvent> {
    return this._onExit.event;
  }

  private async _setup(
    file: string,
    args: string[],
    opts?: IPtySpawnOptions,
  ): Promise<void> {
    // Phase 1: create PTY + spawn child (reader NOT started yet)
    this._pid = await invoke<number>("pty_spawn", {
      file,
      args,
      cols: this.cols,
      rows: this.rows,
      cwd: opts?.cwd ?? null,
      env: opts?.env ?? {},
    });

    // Exit is a low-frequency signal — keep it as a plain event.
    this._unlistenExit = await listen<{ exit_code: number }>(
      `pty:exit:${this._pid}`,
      (event) => {
        this._onExit.fire({ exitCode: event.payload.exit_code });
        this._cleanup();
        // Free the Rust-side session (FDs, memory)
        invoke("pty_close", { pid: this._pid }).catch((err) => {
          terminalLog("pty_close failed:", errorMessage(err));
        });
      },
    );

    // Guard: if kill() was called while setup was in flight, abort
    if (this._destroyed) {
      this._cleanup();
      await invoke("pty_kill", { pid: this._pid }).catch((err) => {
        terminalLog("pty_kill (setup guard) failed:", errorMessage(err));
      });
      // Listeners were torn down, so the exit handler won't free the session
      // (#974). Close it explicitly here too.
      await invoke("pty_close", { pid: this._pid }).catch((err) => {
        terminalLog("pty_close (setup guard) failed:", errorMessage(err));
      });
      return;
    }

    // Binary output Channel. Wiring onmessage before pty_start means the reader
    // cannot send before we are listening — no data-loss race (WI-1.1).
    const channel = new Channel<ArrayBuffer | Uint8Array | number[]>();
    channel.onmessage = (msg) => {
      this._onData.fire(toUint8Array(msg));
    };
    this._dataChannel = channel;

    // Start the reader thread, handing it the channel as `onBytes`.
    try {
      await invoke("pty_start", { pid: this._pid, onBytes: channel });
    } catch (err) {
      this._cleanup();
      throw err;
    }
  }

  write(data: string): void {
    this._ready
      .then(() => invoke("pty_write", { pid: this._pid, data }))
      .catch((err) => {
        ptyWarn("pty_write failed:", errorMessage(err));
      });
  }

  resize(columns: number, rows: number): void {
    this.cols = columns;
    this.rows = rows;
    this._ready
      .then(() =>
        invoke("pty_resize", { pid: this._pid, cols: columns, rows }),
      )
      .catch((err) => {
        ptyWarn("pty_resize failed:", errorMessage(err));
      });
  }

  kill(): void {
    this._destroyed = true;
    this._cleanup();
    // _cleanup() removed the pty:exit listener, so the natural exit handler
    // that calls pty_close never runs (#974). Close the Rust session
    // explicitly after pty_kill — in finally so a failed kill still frees the
    // session map entry (FDs, channels, child handle).
    this._ready
      .then(() => invoke("pty_kill", { pid: this._pid }))
      .catch((err) => {
        terminalLog("pty_kill failed:", errorMessage(err));
      })
      .finally(() => {
        invoke("pty_close", { pid: this._pid }).catch((err) => {
          terminalLog("pty_close failed:", errorMessage(err));
        });
      });
  }

  pause(): void {
    this._ready
      .then(() => invoke("pty_pause", { pid: this._pid }))
      .catch((err) => {
        terminalLog("pty_pause failed:", errorMessage(err));
      });
  }

  resume(): void {
    this._ready
      .then(() => invoke("pty_resume", { pid: this._pid }))
      .catch((err) => {
        terminalLog("pty_resume failed:", errorMessage(err));
      });
  }

  private _cleanup(): void {
    // The Channel has no "unlisten"; drop its onmessage so no further bytes
    // reach the (disposed) consumer, and release the reference.
    if (this._dataChannel) {
      this._dataChannel.onmessage = () => {};
      this._dataChannel = null;
    }
    this._unlistenExit?.();
    this._unlistenExit = null;
  }
}

/** Coerce a Channel payload to a Uint8Array. Raw bodies arrive as ArrayBuffer;
 *  the other branches are defensive (a typed-array view or a JSON number[]). */
function toUint8Array(msg: ArrayBuffer | Uint8Array | number[]): Uint8Array {
  if (msg instanceof Uint8Array) return msg;
  if (msg instanceof ArrayBuffer) return new Uint8Array(msg);
  if (ArrayBuffer.isView(msg)) {
    const view = msg as ArrayBufferView;
    return new Uint8Array(view.buffer, view.byteOffset, view.byteLength);
  }
  return new Uint8Array(msg);
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export function spawn(
  file: string,
  args: string[] | string,
  options?: IPtySpawnOptions,
): IPty {
  const argArray = typeof args === "string" ? [args] : args;
  return new VMarkPty(file, argArray, options);
}

File: src-tauri/src/pty.rs
//! PTY process management — async-safe replacement for tauri-plugin-pty.
//!
//! Purpose: Spawns and manages pseudo-terminal sessions for VMark's built-in
//! terminal. Each PTY reader runs on a dedicated OS thread so blocking I/O
//! never starves the tokio async runtime. Output bytes are pushed to the
//! frontend over a per-session binary `Channel`; the exit signal is a Tauri
//! event. Neither path polls.
//!
//! Key decisions:
//!   - Reader threads are plain `std::thread`, NOT `tokio::spawn_blocking`.
//!     PTY reads are long-lived (lifetime of the shell), so they should not
//!     consume the tokio blocking thread pool.
//!   - Output transport is a `tauri::ipc::Channel<InvokeResponseBody>` (WI-1.1,
//!     ADR-T1): the reader sends `InvokeResponseBody::Raw(bytes)`, delivered to
//!     the webview as a binary `ArrayBuffer` (not a JSON number array) and
//!     point-to-point (no `app.emit` broadcast to every window).
//!   - Pause/resume uses `Condvar` so a paused reader truly sleeps (zero CPU)
//!     instead of busy-waiting.
//!   - Two-phase startup: `pty_spawn` creates the session (on the blocking
//!     pool — `openpty`/`spawn_command` are synchronous syscalls), `pty_start`
//!     begins the reader thread. The frontend wires the output Channel's
//!     `onmessage` and the `pty:exit:{pid}` listener before calling
//!     `pty_start`, so no output or exit signal is lost (no data-loss race).
//!   - Child exit is detected in the reader thread (after the read loop ends)
//!     via `child.wait()`, then emitted as a `pty:exit:{pid}` event.
//!   - Sessions are removed from the map via `pty_close` (called by the
//!     frontend after receiving the exit event) to prevent FD/memory leaks.
//!     A close BETWEEN spawn and start kills + reaps the still-owned child
//!     (no reader thread exists yet to `wait()` on it).
//!   - Writer and master use `std::sync::Mutex` (not tokio) because the
//!     underlying operations are plain syscalls, not async I/O. Writes still
//!     run inside `spawn_blocking`: `write_all` blocks when the PTY buffer is
//!     full (e.g. a large paste into a non-reading foreground process), and a
//!     blocked tokio worker would starve the runtime.
//!
//! @coordinates-with lib.rs — commands registered in generate_handler![]
//! @coordinates-with src/lib/pty.ts — frontend wrapper (output Channel + exit event)
//! @module pty

mod session;

pub use session::{kill_all, PtyState};

use crate::command_error::CommandError;
use portable_pty::PtySize;
use session::{get_session, PtyExitEvent};
use std::collections::BTreeMap;
use std::io::{Read, Write};
use std::sync::atomic::Ordering;
use std::sync::Arc;
use tauri::{AppHandle, Emitter, Runtime};

// WI-DP2.2 — the PTY command surface, typed. Three classes, and the reason each
// is what it is:
//
//   not-found  an unknown pid. The frontend polls these after a `pty:exit`
//              event, so "the session is gone" is an ordinary race, not a fault
//              — it must be distinguishable from a real failure.
//   io         the pty write/resize/kill itself failed. A real device error.
//   internal   a poisoned mutex or a tokio join failure: this process is in a
//              state it should not be able to reach, and no caller can act on
//              it. NOT the catch-all — ADR-2 forbids `internal` as a shortcut
//              for "unclassified", which is why the three above are separate.
fn session_gone(pid: u32) -> CommandError {
    CommandError::not_found(format!("unknown PTY session {pid}"))
}

fn pty_io(error: impl std::fmt::Display) -> CommandError {
    CommandError::io(error.to_string())
}

fn pty_internal(what: &str, error: impl std::fmt::Display) -> CommandError {
    CommandError::internal(format!("{what}: {error}"))
}

// ---------------------------------------------------------------------------
// Commands
// ---------------------------------------------------------------------------

/// Create a PTY session and spawn the child process.
/// Returns the session PID. Call `pty_start` after registering event listeners.
///
/// The blocking pieces (`openpty`, `spawn_command`) run on the blocking pool
/// — see the module header's async-safety rules; only the session-map insert
/// touches the async runtime.
#[tauri::command]
pub async fn pty_spawn(
    file: String,
    args: Vec<String>,
    cols: u16,
    rows: u16,
    cwd: Option<String>,
    env: BTreeMap<String, String>,
    state: tauri::State<'_, PtyState>,
) -> Result<u32, CommandError> {
    let session = tokio::task::spawn_blocking(move || {
        session::create_session(file, args, cols, rows, cwd, env)
    })
    .await
    .map_err(|e| pty_internal("PTY spawn task failed", e))?
    .map_err(pty_io)?;

    let pid = state.next_id.fetch_add(1, Ordering::Relaxed);
    state.sessions.write().await.insert(pid, Arc::new(session));
    Ok(pid)
}

/// Start the reader thread for a PTY session.
/// Must be called exactly once per session, after the caller has wired the
/// `on_bytes` Channel and the `pty:exit:{pid}` listener.
/// Sends output bytes over `on_bytes` (Raw → ArrayBuffer); emits
/// `pty:exit:{pid}` on child exit.
#[tauri::command]
pub async fn pty_start<R: Runtime>(
    pid: u32,
    on_bytes: tauri::ipc::Channel<tauri::ipc::InvokeResponseBody>,
    state: tauri::State<'_, PtyState>,
    app: AppHandle<R>,
) -> Result<(), CommandError> {
    let session = get_session(&state, pid).await.map_err(|_| session_gone(pid))?;
    let mut reader = session
        .reader
        .lock()
        .await
        .take()
        .ok_or_else(|| CommandError::conflict("pty_start already called for this session"))?;
    let mut child = session
        .child
        .lock()
        .await
        .take()
        .ok_or_else(|| CommandError::conflict("pty_start already called for this session"))?;
    let pause_ctl = session.pause_ctl.clone();
    let shutdown = session.shutdown.clone();

    let exit_event = format!("pty:exit:{pid}");

    // Reader thread body wrapped in catch_unwind so a panic doesn't crash
    // the whole process. We must keep the explicit thread name (`pty-reader-{pid}`)
    // for log filtering, so we use Builder directly rather than spawn_thread_logged.
    //
    // `child.wait()` and the exit emit live OUTSIDE the catch_unwind below so
    // they run on BOTH the normal and the (defensive) panic path — the child is
    // always reaped, never a zombie, and the frontend always gets its exit.
    let session_for_panic = session.clone();
    std::thread::Builder::new()
        .name(format!("pty-reader-{pid}"))
        .spawn(move || {
            let pid_for_log = pid;
            let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
                // 64 KB buffer (WI-1.2): far fewer reads/sends per burst than the
                // old 4 KB, which compounds with the binary Channel (WI-1.1).
                let mut buf = vec![0u8; 65536];
                loop {
                    if shutdown.load(Ordering::Acquire) {
                        break;
                    }
                    pause_ctl.wait_if_paused();
                    if shutdown.load(Ordering::Acquire) {
                        break;
                    }
                    match reader.read(&mut buf) {
                        Ok(0) => break,
                        Ok(n) => {
                            // Send raw bytes over the per-session Channel. Raw →
                            // ArrayBuffer in the webview (WI-1.1), point-to-point
                            // (no app.emit broadcast → closes T2). A send error
                            // means the webview/channel is gone; kill the child so
                            // the child.wait() below returns instead of hanging on
                            // a still-running shell, then stop reading.
                            if on_bytes
                                .send(tauri::ipc::InvokeResponseBody::Raw(buf[..n].to_vec()))
                                .is_err()
                            {
                                if let Ok(mut killer) = session_for_panic.child_killer.lock() {
                                    let _ = killer.kill();
                                }
                                break;
                            }
                        }
                        // EINTR is transient (signal during read) — retry rather
                        // than treat it as EOF and prematurely end the session.
                        Err(ref e) if e.kind() == std::io::ErrorKind::Interrupted => continue,
                        // Any other read error ends the session. Log it before
                        // breaking so a terminal that dies on a read error leaves
                        // a diagnostic (WI-4.3 / G8) instead of vanishing silently.
                        // Kill the child too: on a fatal read error the shell may
                        // still be running, and the child.wait() below would then
                        // block this thread forever instead of reaping.
                        Err(e) => {
                            log::warn!(
                                "[pty] reader {pid_for_log} read error ({:?}): {e}",
                                e.kind(),
                            );
                            if let Ok(mut killer) = session_for_panic.child_killer.lock() {
                                let _ = killer.kill();
                            }
                            break;
                        }
                    }
                }
            }));
            // On panic the read loop bailed early and the shell may still be
            // running — kill it so the wait() below returns instead of blocking.
            // (The loop has no panic-able operation; this is defense in depth.)
            if let Err(payload) = result {
                log::error!(
                    "[task:pty-reader-{}] reader thread panicked: {}",
                    pid_for_log,
                    crate::task::panic_payload_message(&payload),
                );
                if let Ok(mut killer) = session_for_panic.child_killer.lock() {
                    let _ = killer.kill();
                }
            }
            // Reap the child and emit exit on BOTH the normal and panic paths —
            // guarantees no zombie even if the reader panicked (audit follow-up).
            let exit_code = child.wait().map(|s| s.exit_code()).unwrap_or(1);
            let _ = app.emit(&exit_event, PtyExitEvent { exit_code });
        })
        .map_err(|e| pty_internal("failed to spawn PTY reader thread", e))?;

    Ok(())
}

/// Write data to the PTY.
///
/// The write runs on the blocking pool: a full PTY buffer makes `write_all`
/// block until the foreground process reads, which would otherwise pin a
/// tokio worker thread (and the writer mutex) for the duration.
#[tauri::command]
pub async fn pty_write(
    pid: u32,
    data: String,
    state: tauri::State<'_, PtyState>,
) -> Result<(), CommandError> {
    let session = get_session(&state, pid).await.map_err(|_| session_gone(pid))?;
    tokio::task::spawn_blocking(move || {
        let mut writer = session
            .writer
            .lock()
            .map_err(|e| pty_internal("PTY writer lock poisoned", e))?;
        writer.write_all(data.as_bytes()).map_err(pty_io)?;
        writer.flush().map_err(pty_io)
    })
    .await
    .map_err(|e| pty_internal("PTY write task failed", e))?
}

/// Resize the PTY.
#[tauri::command]
pub async fn pty_resize(
    pid: u32,
    cols: u16,
    rows: u16,
    state: tauri::State<'_, PtyState>,
) -> Result<(), CommandError> {
    let session = get_session(&state, pid).await.map_err(|_| session_gone(pid))?;
    let master = session
        .master
        .lock()
        .map_err(|e| pty_internal("PTY master lock poisoned", e))?;
    master
        .resize(PtySize {
            rows: rows.max(1),
            cols: cols.max(1),
            pixel_width: 0,
            pixel_height: 0,
        })
        .map_err(pty_io)
}

/// Kill the PTY child process.
#[tauri::command]
pub async fn pty_kill(pid: u32, state: tauri::State<'_, PtyState>) -> Result<(), CommandError> {
    let session = get_session(&state, pid).await.map_err(|_| session_gone(pid))?;
    session.shutdown.store(true, Ordering::Release);
    session.pause_ctl.resume(); // Wake reader if paused
    let mut killer = session
        .child_killer
        .lock()
        .map_err(|e| pty_internal("PTY child-killer lock poisoned", e))?;
    killer.kill().map_err(pty_io)
}

/// Remove session from the map, freeing FDs and memory.
/// Called by the frontend after receiving the `pty:exit:{pid}` event — and
/// also legal between `pty_spawn` and `pty_start`, where the session still
/// owns the child: no reader thread exists to reap it, so close kills and
/// reaps it here (a bare map-remove would drop the child unreaped).
#[tauri::command]
pub async fn pty_close(pid: u32, state: tauri::State<'_, PtyState>) -> Result<(), CommandError> {
    let Some(session) = state.sessions.write().await.remove(&pid) else {
        return Ok(());
    };
    tokio::task::spawn_blocking(move || session::kill_and_reap_unstarted(&session))
        .await
        .map_err(|e| pty_internal("PTY close task failed", e))
}

/// Pause the PTY reader (flow control).
#[tauri::command]
pub async fn pty_pause(pid: u32, state: tauri::State<'_, PtyState>) -> Result<(), CommandError> {
    let session = get_session(&state, pid).await.map_err(|_| session_gone(pid))?;
    session.pause_ctl.pause();
    Ok(())
}

/// Resume the PTY reader (flow control).
#[tauri::command]
pub async fn pty_resume(pid: u32, state: tauri::State<'_, PtyState>) -> Result<(), CommandError> {
    let session = get_session(&state, pid).await.map_err(|_| session_gone(pid))?;
    session.pause_ctl.resume();
    Ok(())
}
